Analyzer/datadog

amanfcp opened the pull request:
### Description:
This PR adds analyzer for DataDog
DataDog API docs: [Link](https://docs.datadoghq.com/api/latest/)


### Checklist:
* [x] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?

amanfcp committed:
task: Datadog analyzer init commit with limited resources and permissions
No cases covered

amanfcp committed:
Merge branch 'main' of amaandev:amanfcp/trufflehog into analyzer/datadog

amanfcp committed:
task: added more permissions

nabeelalam commented on pkg/analyzer/analyzers/datadog/requests.go:
```diff
@@ -0,0 +1,321 @@
+		}
+		return nil
+	case http.StatusUnauthorized, http.StatusForbidden:
+		return nil
```
This is in draft and may have been updated, but in `http.StatusUnauthorized || http.StatusForbidden` case, is there a reason not to return an authorization error?

amanfcp committed:
task: added more permissions | refactoring

amanfcp committed:
Merge branch 'main' of amaandev:amanfcp/trufflehog into analyzer/datadog

amanfcp committed:
task: added domain handling

amanfcp commented on pkg/analyzer/analyzers/datadog/datadog.go:
```diff
@@ -0,0 +1,173 @@
+		binding := analyzers.Binding{
+			Resource: *secretInfoResourceToAnalyzerResource(resource),
+			Permission: analyzers.Permission{
+				Value: "admin", // Using admin as default permission level
```
This is a work in progress. Will surely cater this. Thanks for the suggestions @shahzadhaider1 :100:

amanfcp commented on pkg/analyzer/analyzers/datadog/requests.go:
```diff
@@ -0,0 +1,315 @@
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	// create request
+	req, err := http.NewRequestWithContext(ctx, method, baseURL+endpoint, http.NoBody)
+	if err != nil {
+		return nil, 0, err
+	}
```
Yes, `POST`, `PUT` and `PATCH` methods are being used. The use case is to only verify if the user is able to access the endpoint. Hence there is no need to implement sending dynamic `body`.

amanfcp committed:
refactor: incorporated PR feedback

amanfcp committed:
Merge branch 'main' into analyzer/datadog

amanfcp committed:
task: Analyzer Result implementation for enterprise

amanfcp committed:
task: added tests | Analysis Info configuration changes | permissions.yaml

amanfcp committed:
revert: reverted datadog detector base URL

amanfcp committed:
Merge branch 'main' of github.com:trufflesecurity/trufflehog into analyzer/datadog

amanfcp commented on pkg/analyzer/analyzers/datadog/datadog.go:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	// Get appKey if provided
+	appKey := credInfo["appKey"]
```
We can verify only the `apiKey` through `/validate` endpoint even if the `appKey` is not provided. Following the same flow as detector.

amanfcp commented on pkg/analyzer/analyzers/datadog/datadog.go:
```diff
@@ -0,0 +1,219 @@
+		userResource = &analyzers.Resource{
+			Name: info.User.Name,
+			Type: "user",
+			Metadata: map[string]any{
+				"email": info.User.Email,
+			},
```
Resolving it as per the discussion with Dustin.

amanfcp commented on pkg/analyzer/analyzers/datadog/datadog.go:
```diff
@@ -0,0 +1,219 @@
+
+	// Create user resource to use as parent
+	var userResource *analyzers.Resource
+	if info.User.Id != "" {
```
As per the `SecretInfo` struct,  `info.User` will be assigned zero values.

amanfcp committed:
refactor: streamline Datadog API key analysis and permission handling

- Simplified API key retrieval logic by removing redundant checks for "key".
- Enhanced permission binding extraction by introducing a new function for permissions.
- Cleaned up resource binding logic for better clarity and maintainability.
- Removed unnecessary ID fields from scopes.json to reduce clutter.

amanfcp committed:
Merge branch 'main' into analyzer/datadog

amanfcp committed:
Merge branch 'main' of github.com:trufflesecurity/trufflehog into analyzer/datadog

amanfcp committed:
Merge branch 'main' into analyzer/datadog

amanfcp committed:
Merge branch 'main' into analyzer/datadog
